Recheck for URLs whenever terminal content changes, as well as other events

CrackedP0t opened the pull request:
This PR makes it so that whenever the terminal's dirty, as well as on certain events, Alacritty checks for URLs - but only if the mouse is in the window.

I changed a good bit of code, so critique is welcome!

Should fix #2388

chrisduerr commented:
I've run this through `perf` with a couple of different variations which I believe should stress this quite a lot and I was not able to find any problems where it would show up in the significant parts of the benchmark results.

So I'd focus on making sure that the underline does not flicker and once that's done rerun the benchmarks again.

CrackedP0t commented:
Thanks for profiling this!
We could do more extensive checking to see if a re-highlight's really necessary, but I don't think it would be worth it. Even the worst case of iterating over every character every frame, while not trivial, doesn't appear to me to be very impactful when compared to other routines in Alacritty.
I'll look into the flicker!

chrisduerr commented:
Yeah, it looks like it was the correct choice to parse inside-out from the cursor position since that usually pretty cheap. Even in bad cases it's still only done once per frame though, other operations which are done for every single cell seem much more intense.

The primary focus right now should be the flickering, I think that's an important one to get out of the way first, let me know if you need help with reproducing it, hopefully my command should work.

There might be some value in further benchmarking though. I've never tested with the absolute worst-case of basically a single URL spanning the complete history and frequent redraws. So that could be tested when the flickering is resolved. Parsing 10_000 lines x width of the terminal every frame would likely cause some troubles even though it's not a realistic scenario.

CrackedP0t commented:
Alright, I fixed the flickering issue - it was caused by the main loop dropping `terminal_lock` between examining the terminal and displaying it. If input was arriving fast enough, the terminal would attempt to draw a highlight over the wrong text. I changed `display.draw` to take a mutable reference instead of a `FairMutex`, reusing the original lock.

I also set a limit of 2000 characters on searching for a URL - that seems to be the generally-accepted maximum length that most programs will deal with. I can't imagine that anyone will actually want to click on and open a URL longer than that in practical use.

CrackedP0t commented on alacritty_terminal/src/input.rs:
```diff
@@ -425,7 +421,7 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
         if cell_changed {
-            self.update_url_highlight(point, modifiers);
+            self.ctx.terminal_mut().url_dirty = true;
         }
```
Good point

CrackedP0t commented on alacritty/src/main.rs:
```diff
@@ -243,6 +243,12 @@ fn run(
+            processor.update_url_highlight(&mut terminal_lock);
+        }
+
+        terminal_lock.url_dirty = false;
```
Well, one case in which `url_dirty` will stay `true` into the next frame if it's in the `if` block is when `url_dirty` is set when the mouse is outside the window.

CrackedP0t commented on alacritty_terminal/src/event.rs:
```diff
@@ -455,6 +463,15 @@ impl<N: Notify> Processor<N> {
+                    },
+                    CursorLeft { .. } => {
+                        processor.ctx.mouse.inside_window = false;
+                    }
```
In what way, sorry?

chrisduerr replied:
If there's only a single line, it should be formatted without dedicated block.

CrackedP0t replied:
Oh, that makes sense, thanks!

CrackedP0t commented:
That's a good point - I'll benchmark to see what a reasonable limit would be for URL length.

CrackedP0t commented:
Alright - when searching through 25000 characters, on my computer the time for `url_check` to complete was between 19 and 6 milliseconds, and I didn't feel much or any lag when typing. I'd definitely be open to discussing the limit though.

cole-h commented on alacritty_terminal/src/term/mod.rs:
```diff
@@ -47,7 +47,7 @@ pub mod cell;
 pub mod color;
 
 // De facto limit - see https://stackoverflow.com/a/417184/
```
Small nit: since you're changing the max length away from  the "de facto" proposed in that overflow answer, would it be a good idea to remove this comment?

CrackedP0t replied:
Oh, yep!

chrisduerr commented:
> 25000 characters seems to not lag too much; will only rarely hit this anyways

I don't think 'not lag too much' lives up to the performance standard of Alacritty, neither does 'rarely hit this anyways' seem to be a good excuse (can't really tell people that Alacritty is only fast in 90% of the cases, it's either fast or it's not).

However I do understand that this creates quite the conundrum. Long URLs are problematic for performance, however short URLs are bad for the user since URLs potentially won't open at all. A relatively standard maximum URL length like 65536 is too high and a relatively arbitrary lower limit of 2000 will definitely cut up some URLs.

All of this is why I generally did not want to run URL detection on every update. The way it is currently on the latest master might be a bit limited but it allows us to pretty much circumvent all performance pitfalls (the only one a case could be made for is on a very long URL when the mouse is moved repeatedly over a huge potential URL).

Now that we've established that there is a problem, the question is how to best solve this problem. I think one path forwards should definitely be to test what URL lengths other terminal emulators support. I know that URxvt will support arbitrary URL length (I think it's until the start leaves the scrollback) and it would be unfortunate to 'lose' against it here, but maybe other terminals also make this tradeoff of limiting URL length?

Another solution would be to try and clear URL highlights more regularly, but not redrawing them unless at least the mouse moves or the modifiers change. I don't think it's super necessary to update URL highlights when the text changes under a stationary cursor, however clearing appropriately would probably also be tough to do properly.

In general we also need a way to detect URLs without the mouse anyways, since we want to eventually make it possible to have key bindings to open links and I've actually written a URL parser which would be more appropriate for that (it operates on a stream rather than a buffer), so maybe exploring into that direction is another option. Though right now that is not necessary yet so it's probably not going to help.

I haven't looked into the performance of the latest version of this PR myself yet, however I see a great risk in just rushing this change out especially because it affects some of the most essential performance areas for high throughput. It would be really unfortunate if Alacritty turns out to be mediocre at playing ascii videos because it turns out it things everything's a URL.

CrackedP0t commented:
Well, if you're planning on reworking the URL detection anyway, I think it might make sense to either merge this PR with a lower length limit or wait to merge it until you've pushed the URL rework. I do think that this PR is better than the current code, and it would make sense to merge it - however, I'd be open to any discussion

nixpulvis commented:
Just taking a look at this now hoping it would fix a problem I'm having with the following config:

```yaml
mouse:
  url:
    modifiers: Control
```

The issue is that the `MouseCursor::Hand` and underline aren't showing up when already hovering over a URL and pressing (in this case) <kbd>Control</kbd>. On master the styles change if the hover happens _while_ the modifier is held. However, on this branch the style never changes.

chrisduerr commented:
I've published my library for finding URLs in reverse character streams [here](https://crates.io/crates/rfind_url) and while the latest changes for bracket and quote matching logic have significantly slowed things down (I haven't had the time to look into optimizations yet), I think it might allow us to take a better approach for URL searching in Alacritty.

One of the main goals here of course is not to improve performance, but to make it possible to detect multiple URLs at arbitrary positions for opening them using keyboard shortcuts. However, I think this should still fix problems with searching large buffers just for detecting that something is not a URL, since the library does not make any allocations (except for the bracket logic), opposed to the current implementation which stores the complete potential match in a String. So that should resolve the remaining troubles I've had with this PR. Putting an upper bound on things would probably still make sense, but we could probably bump that above anything reasonable.

Of course I do not expect you to figure out how to implement this crate in a reasonable manner so I'll try to hook it up to Alacritty myself. I'm not exactly sure yet what the best starting point might be, but I'll update this issue once I've figured something out. I'll still leave this open for now since it might be useful to base this work on top of this PR or it might turn out to be more difficult than expected, but hopefully we'll be able to figure something out for this soon.

kchibisov commented:
FYI, the `rfind_url` rework and other `url` refactorings were merged.

chrisduerr commented:
Good point @kchibisov. Please note that I've checked if the `rfind_url` branch fixes the DoS problem with huge URLs and it has improved the situation, however simply bumping the URL length by a factor of about 10 made it just as bad again. So with a scrollback length of 100_000 lines there's absolutely no way we can completely sidestep this issue by optimizing performance and we will have to introduce some limits.

I still don't think updating the URL state whenever the content of the terminal changes is a good idea. That's just going to create more troubles than it's worth. I think it would make a lot more sense to track rotation and clearing and handle it in a similar way to selection (though selection currently doesn't properly track rotation in scrolling regions).

nixpulvis commented:
This is somewhat related to #2719.

CrackedP0t commented:
Sure, I'd be happy to look into implementing this - I think it sounds like a good idea to track and remember URLs.
